p2p/src/supervisor.rs

```diff
@@ -0,0 +1,517 @@
+    /// like the number of concurrently connected peers should issue a disconnect to remedy the
+    /// situation.
+    Accept,
+    /// Establishes a new connection to the remote end in [`transport::ConnectInfo`].
+    Connect(transport::ConnectInfo),
+    /// Disconnects the [`peer::Peer`] known by [`node::Id`]. This will tear down the entire tree of
+    /// subroutines managing the peer in question.
+    Disconnect(node::Id),
+    /// Dispatch the given message to the peer known for [`node::Id`].
+    Msg(node::Id, message::Send),
+}
+
+/// Set of significant events in the p2p subsystem.
+pub enum Event {
+    /// A new connection has been established.
+    Connected(node::Id, Direction),
+    /// A [`peer::Peer`] has been disconnected.
+    Disconnected(node::Id, Report),
+    /// A new [`message::Receive`] from the [`peer::Peer`] has arrived.
+    Message(node::Id, message::Receive),
+    /// The [`Supervisor`] has terminated and should be dropped immediatelly. Optionally carrying
+    /// an error which caused the termination.
+    Terminated(Option<Report>),
+    /// A connection upgraded successfully to a [`peer::Peer`].
+    Upgraded(node::Id),
+    /// An upgrade from failed.
+    UpgradeFailed(node::Id, Report),
+}
+
+/// Fatal errors from the supervisor machinery.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// Error during subroutine execution.
+    #[error("subroutine execution")]
+    Join(String, Report),
+    /// Internal state lock is poisoned and can't be recovered.
+    #[error("state lock poisoned")]
+    StateLockPoisoned,
+}
+
+/// Wrapping a [`transport::Transport`] the `Supervisor` runs the p2p machinery to manage peers over
+/// physical connections. Offering multiplexing of ingress and egress messages and a surface to
+/// empower higher-level protocols to control the behaviour of the p2p substack.
+///
+/// TODO(xla): Document subroutine/thread hierarchy and data flow.
+pub struct Supervisor {
+    command_tx: Sender<Command>,
+    event_rx: Receiver<Event>,
+}
+
+impl Supervisor {
+    /// Takes the [`transport::Transport`] and sets up managed subroutines. When the `Supervisor`
+    /// is returned the p2p subsystem has been successfully set up on the given network interface
+    /// (as far as applicable for the transport) so the caller can use the command input and
+    /// consume events.
+    ///
+    /// # Errors
+    ///
+    /// * If the bind of the transport fails
+    pub fn run<T>(transport: T, info: transport::BindInfo) -> Result<Self>
+    where
+        T: transport::Transport + Send + 'static,
+    {
+        let (endpoint, incoming) = transport.bind(info)?;
+        let (command_tx, command_rx) = unbounded();
+        let (event_tx, event_rx) = unbounded();
+
+        thread::spawn(move || {
+            match Self::main::<T>(command_rx, event_tx.clone(), endpoint, incoming) {
+                Ok(handles) => {
+                    for (_name, handle) in handles {
```

Comment:
> Does that ordering matter?

Ordering does not matter, unless you are seeing a case where it would.

> What happens when one of the threads beyond the `supervisor-accept` (first) thread fail? e.g. if the `supervisor-message` thread fails, but the `supervisor-accept` thread continues running, won't this effectively leave the supervisor in a partially degraded/invalid state of operation, where it can accept incoming peer connections but can't handle messages from them?

All subroutines have two channels, one which they read from to initiate the next operation (including accept) and one to send the result of the operation. If one of them fails the supervisor run loop is terminated - which in turn means that all Senders are dropped and therefore the receive errors. Symmetrically if one of the subroutiens fails (the code is crafted in a way that all side-effects are accounted for so panics in subroutines should be eliminated) the Sender will error on the next operation. With this approach there will be convergence onto the actual state of affairs, there is room for a single in-flight operation to be underway before the machinery is dismantled, which in my view is a worthwhile trade-off as long as the entire system converges towards the true state. Super important to surface if we see scenarios where this can leave the p2p stack in a partially degraded state.